<!DOCTYPE html>
<html>
<head>

<meta charset='utf-8'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, user-scalable=no'>
<meta name='apple-touch-fullscreen'       content='yes'>
<meta name='apple-mobile-web-app-capable' content='yes'>
<meta name='apple-mobile-web-app-status-bar-style' content='rgba(228,228,228,1.0)'>

<title>Exception: Exception &mdash; Core  Ruby-3.3.9 p170</title>

<link rel='stylesheet'  type='text/css' href='../../css/y_fonts.css' />
<link rel='stylesheet'  type='text/css' href='../../css/highlight.github.css' />
<link rel='stylesheet'  type='text/css' href='../../css/y_style.css' />
<link rel='stylesheet'  type='text/css' href='../../css/y_list.css' />
<link rel='stylesheet'  type='text/css' href='../../css/y_color.css' />

<script type='text/javascript'>
  var pathId = "Exception",
    relpath = '';

  var t2Info = {
    CSEP: '.',
    ISEP: '#',
    NSEP: '::'
  };
</script>

<script type='text/javascript' charset='utf-8' src='../../js/highlight.pack.js'></script>
<script type='text/javascript' charset='utf-8' src='../../js/y_app.js'></script>

</head>
<body>
<svg id='y_wait' class viewBox='0 0 90 90'></svg>
<div id='settings' class='hidden'></div>
<div id='y_list' class='d h'>
  <header id='list_header'></header>
  <nav id= 'list_nav' class='y_nav l_nav'>
    <ul id='list_items'></ul>
  </nav>
</div>
<div id='y_toc' class='f h'>
  <header id='toc_header'></header>
  <nav id= 'toc_nav' class='y_nav t_nav'>
  <ol id='toc_items'></ol>
  </nav>
</div>
<div id='y_main' tabindex='-1'>
  <header id='y_header'>
    <div id='y_menu'>
      <a id='home_no_xhr' href='/'>Home</a> &raquo; 
      <a href='../.'>Ruby-3.3.9</a> &raquo; 
      <a href='.'>Core</a> &raquo; 
      <a href='_index.html#alpha_E'>Index (E)</a> &raquo; 
      <span class='title'><a id='t2_doc_top' href='#'>Exception&nbsp;&#x25B2;</a></span>
    </div>

    <a id='list_href' href="class_list.html"></a>
    <div id='y_measure_em' class='y_measure'></div>
    <div id='y_measure_vh' class='y_measure'></div>
    <span id='y_measure_50pre' class='y_measure'><code>123456789_123456789_123456789_123456789_123456789_</code></span>
  </header>
<div id='content' class='class'>
<h1>Exception: Exception</h1>
<div id='t2_rel_sf_div'>
<table id='t2_rel_sf' class='y_box'>
  <thead>
    <tr><td id='t2_relations' colspan='2'>Relationships & Source Files</td></tr>
  </thead>
  <tbody>
    <tr><td class='box_h' colspan='2'>Extension / Inclusion / Inheritance Descendants</td></tr>
    <tr>
      <td colspan='2'>
        <div class='box_2'>Subclasses:</div>
        <div class='box_22'>
          <a href="ArgumentError.html" title="ArgumentError (class)">ArgumentError</a>, <a href="ClosedQueueError.html" title="ClosedQueueError (class)">ClosedQueueError</a>, <a href="EOFError.html" title="EOFError (class)">EOFError</a>, <a href="EncodingError.html" title="EncodingError (class)">EncodingError</a>, <a href="FiberError.html" title="FiberError (class)">FiberError</a>, <a href="FloatDomainError.html" title="FloatDomainError (class)">FloatDomainError</a>, <a href="FrozenError.html" title="FrozenError (class)">FrozenError</a>, <a href="IOError.html" title="IOError (class)">IOError</a>, <a href="IndexError.html" title="IndexError (class)">IndexError</a>, <a href="Interrupt.html" title="Interrupt (class)">Interrupt</a>, <a href="KeyError.html" title="KeyError (class)">KeyError</a>, <a href="LoadError.html" title="LoadError (class)">LoadError</a>, <a href="LocalJumpError.html" title="LocalJumpError (class)">LocalJumpError</a>, <a href="NameError.html" title="NameError (class)">NameError</a>, <a href="NoMatchingPatternError.html" title="NoMatchingPatternError (class)">NoMatchingPatternError</a>, <a href="NoMatchingPatternKeyError.html" title="NoMatchingPatternKeyError (class)">NoMatchingPatternKeyError</a>, <a href="NoMemoryError.html" title="NoMemoryError (class)">NoMemoryError</a>, <a href="NoMethodError.html" title="NoMethodError (class)">NoMethodError</a>, <a href="NotImplementedError.html" title="NotImplementedError (class)">NotImplementedError</a>, <a href="RangeError.html" title="RangeError (class)">RangeError</a>, <a href="RegexpError.html" title="RegexpError (class)">RegexpError</a>, <a href="RuntimeError.html" title="RuntimeError (class)">RuntimeError</a>, <a href="ScriptError.html" title="ScriptError (class)">ScriptError</a>, <a href="SecurityError.html" title="SecurityError (class)">SecurityError</a>, <a href="SignalException.html" title="SignalException (class)">SignalException</a>, <a href="StandardError.html" title="StandardError (class)">StandardError</a>, <a href="StopIteration.html" title="StopIteration (class)">StopIteration</a>, <a href="SyntaxError.html" title="SyntaxError (class)">SyntaxError</a>, <a href="SystemCallError.html" title="SystemCallError (class)">SystemCallError</a>, <a href="SystemExit.html" title="SystemExit (class)">SystemExit</a>, <a href="SystemStackError.html" title="SystemStackError (class)">SystemStackError</a>, <a href="ThreadError.html" title="ThreadError (class)">ThreadError</a>, <a href="TypeError.html" title="TypeError (class)">TypeError</a>, <a href="UncaughtThrowError.html" title="UncaughtThrowError (class)">UncaughtThrowError</a>, <a href="ZeroDivisionError.html" title="ZeroDivisionError (class)">ZeroDivisionError</a>, <a href="fatal.html" title="fatal (class)">fatal</a>, <a href="Encoding/CompatibilityError.html" title="Encoding::CompatibilityError (class)">Encoding::CompatibilityError</a>, <a href="Encoding/ConverterNotFoundError.html" title="Encoding::ConverterNotFoundError (class)">Encoding::ConverterNotFoundError</a>, <a href="Encoding/InvalidByteSequenceError.html" title="Encoding::InvalidByteSequenceError (class)">Encoding::InvalidByteSequenceError</a>, <a href="Encoding/UndefinedConversionError.html" title="Encoding::UndefinedConversionError (class)">Encoding::UndefinedConversionError</a>, <a href="IO/TimeoutError.html" title="IO::TimeoutError (class)">IO::TimeoutError</a>, <a href="Math/DomainError.html" title="Math::DomainError (class)">Math::DomainError</a>, <a href="Ractor/ClosedError.html" title="Ractor::ClosedError (class)">Ractor::ClosedError</a>, <a href="Ractor/Error.html" title="Ractor::Error (class)">Ractor::Error</a>, <a href="Ractor/IsolationError.html" title="Ractor::IsolationError (class)">Ractor::IsolationError</a>, <a href="Ractor/MovedError.html" title="Ractor::MovedError (class)">Ractor::MovedError</a>, <a href="Ractor/RemoteError.html" title="Ractor::RemoteError (class)">Ractor::RemoteError</a>, <a href="Ractor/UnsafeError.html" title="Ractor::UnsafeError (class)">Ractor::UnsafeError</a>, <a href="Regexp/TimeoutError.html" title="Regexp::TimeoutError (class)">Regexp::TimeoutError</a>, <a href="IO/Buffer/AccessError.html" title="IO::Buffer::AccessError (class)">IO::Buffer::AccessError</a>, <a href="IO/Buffer/AllocationError.html" title="IO::Buffer::AllocationError (class)">IO::Buffer::AllocationError</a>, <a href="IO/Buffer/InvalidatedError.html" title="IO::Buffer::InvalidatedError (class)">IO::Buffer::InvalidatedError</a>, <a href="IO/Buffer/LockedError.html" title="IO::Buffer::LockedError (class)">IO::Buffer::LockedError</a>, <a href="IO/Buffer/MaskError.html" title="IO::Buffer::MaskError (class)">IO::Buffer::MaskError</a>
        </div>
      </td>
    </tr>
    <tr>
      <td id='t2_inherits' class='box_2'>Inherits:</td>
      <td class='box_rel'>
        <span class='inheritName'><a href="Object.html" title="Object (class)">Object</a></span>
      </td>
    </tr>
    <tr class='single'>
      <td id='t2_defined_in' class='box_2'>Defined in:</td>
      <td class='box_rel'><a class='repo' href='https://github.com/ruby/ruby/blob/v3_3_9/error.c#L3146'>error.c</a>      </td>
    </tr>
  </tbody>
</table>
</div>
<h2 id='h2_overview'>Overview</h2>
<div id='overview_ds'>
<div class='docstring'>
  <div class='discussion'>
    
<p>Class <code>Exception</code> and its subclasses are used to communicate between <a href="Kernel.html#raise-class_method" title="Kernel.raise (method)">Kernel.raise</a> and <code>rescue</code> statements in <code>begin ... end</code> blocks.</p>

<p>An Exception object carries information about an exception:</p>
<ul><li>
<p>Its type (the exception’s class).</p>
</li><li>
<p>An optional descriptive message.</p>
</li><li>
<p>Optional backtrace information.</p>
</li></ul>

<p>Some built-in subclasses of <code>Exception</code> have additional methods: e.g., <a href="NameError.html#name-instance_method" title="NameError#name (method)">NameError#name</a>.</p>

<h3 id="label-Defaults">Defaults</h3>

<p>Two Ruby statements have default exception classes:</p>
<ul><li>
<p><code>raise</code>: defaults to <a href="RuntimeError.html" title="RuntimeError (class)"><code>::RuntimeError</code></a>.</p>
</li><li>
<p><code>rescue</code>: defaults to <a href="StandardError.html" title="StandardError (class)"><code>::StandardError</code></a>.</p>
</li></ul>

<h3 id="label-Global+Variables">Global Variables</h3>

<p>When an exception has been raised but not yet handled (in <code>rescue</code>, <code>ensure</code>, <code>at_exit</code> and <code>END</code> blocks), two global variables are set:</p>
<ul><li>
<p><code>$!</code> contains the current exception.</p>
</li><li>
<p><code>$@</code> contains its backtrace.</p>
</li></ul>

<h3 id="label-Custom+Exceptions">Custom Exceptions</h3>

<p>To provide additional or alternate information, a program may create custom exception classes that derive from the built-in exception classes.</p>

<p>A good practice is for a library to create a single “generic” exception class (typically a subclass of <a href="StandardError.html" title="StandardError (class)"><code>::StandardError</code></a> or <a href="RuntimeError.html" title="RuntimeError (class)"><code>::RuntimeError</code></a>) and have its other exception classes derive from that class. This allows the user to rescue the generic exception, thus catching all exceptions the library may raise even if future versions of the library add new exception subclasses.</p>

<p>For example:</p>

<pre class="code ruby"><code class="ruby"><span class='kw'>class</span> <span class='const'>MyLibrary</span>
  <span class='kw'>class</span> <span class='const'>Error</span> <span class='op'>&lt;</span> <span class='op'>::</span><span class='const'><a href="StandardError.html" title="StandardError (class)">StandardError</a></span>
  <span class='kw'>end</span>

  <span class='kw'>class</span> <span class='const'>WidgetError</span> <span class='op'>&lt;</span> <span class='const'>Error</span>
  <span class='kw'>end</span>

  <span class='kw'>class</span> <span class='const'>FrobError</span> <span class='op'>&lt;</span> <span class='const'>Error</span>
  <span class='kw'>end</span>

<span class='kw'>end</span></code></pre>

<p>To handle both <code>MyLibrary::WidgetError</code> and <code>MyLibrary::FrobError</code> the library user can rescue <code>MyLibrary::Error</code>.</p>

<h3 id="label-Built-In+Exception+Classes">Built-In Exception Classes</h3>

<p>The built-in subclasses of <code>Exception</code> are:</p>
<ul><li>
<p><a href="NoMemoryError.html" title="NoMemoryError (class)"><code>::NoMemoryError</code></a></p>
</li><li>
<p><a href="ScriptError.html" title="ScriptError (class)"><code>::ScriptError</code></a></p>
<ul><li>
<p>LoadError</p>
</li><li>
<p>NotImplementedError</p>
</li><li>
<p>SyntaxError</p>
</li></ul>
</li><li>
<p><a href="SecurityError.html" title="SecurityError (class)"><code>::SecurityError</code></a></p>
</li><li>
<p><a href="SignalException.html" title="SignalException (class)"><code>::SignalException</code></a></p>
<ul><li>
<p>Interrupt</p>
</li></ul>
</li><li>
<p><a href="StandardError.html" title="StandardError (class)"><code>::StandardError</code></a></p>
<ul><li>
<p>ArgumentError</p>
<ul><li>
<p>UncaughtThrowError</p>
</li></ul>
</li><li>
<p>EncodingError</p>
</li><li>
<p>FiberError</p>
</li><li>
<p>IOError</p>
<ul><li>
<p>EOFError</p>
</li></ul>
</li><li>
<p>IndexError</p>
<ul><li>
<p>KeyError</p>
</li><li>
<p>StopIteration</p>
<ul><li>
<p>ClosedQueueError</p>
</li></ul>
</li></ul>
</li><li>
<p>LocalJumpError</p>
</li><li>
<p>NameError</p>
<ul><li>
<p>NoMethodError</p>
</li></ul>
</li><li>
<p>RangeError</p>
<ul><li>
<p>FloatDomainError</p>
</li></ul>
</li><li>
<p>RegexpError</p>
</li><li>
<p>RuntimeError</p>
<ul><li>
<p>FrozenError</p>
</li></ul>
</li><li>
<p>SystemCallError</p>
<ul><li>
<p>Errno::*</p>
</li></ul>
</li><li>
<p>ThreadError</p>
</li><li>
<p>TypeError</p>
</li><li>
<p>ZeroDivisionError</p>
</li></ul>
</li><li>
<p><a href="SystemExit.html" title="SystemExit (class)"><code>::SystemExit</code></a></p>
</li><li>
<p><a href="SystemStackError.html" title="SystemStackError (class)"><code>::SystemStackError</code></a></p>
</li><li>
<p>fatal</p>
</li></ul>

  </div>
</div>
</div>
<h2 class='h2_sum' id='class_attribute_summary'>Class Attribute Summary</h2>
<div class='div_sum'>
<ul class='summary full'>
  <li>
    <span class='summary_signature ro'>
      <a href="#to_tty%3F-class_method" title="to_tty? (class method)">.<strong>to_tty?</strong>  &#x21d2; Boolean </a>
    </span>
    <span class='note title readonly'>readonly</span>
    <div class='summary_desc'>
      <div class='inline'><p>Returns <code>true</code> if exception messages will be sent to a tty.</p></div>
    </div>
  </li>
</ul>
</div>  <!-- class_attribute_summary -->

<h2 class='h2_sum' id='class_method_summary'>Class Method Summary</h2>
<div class='div_sum'>
<ul class='summary full'>
  <li>
    <span class='summary_signature '>
      <a href="#exception-class_method" title="exception (class method)">.<strong>exception</strong>([string])  &#x21d2; Exception </a>
    </span>
    <div class='summary_desc'>
      <div class='inline'><p>With no argument, or if the argument is the same as the receiver, return the receiver.</p></div>
    </div>
  </li>
  <li>
    <span class='summary_signature '>
      <a href="#new-class_method" title="new (class method)">.<strong>new</strong>(msg = nil)  &#x21d2; Exception </a>
    </span>
    <span class='note title constructor'>constructor</span>
    <div class='summary_desc'>
      <div class='inline'><p>Construct a new <code>Exception</code> object, optionally passing in a message.</p></div>
    </div>
  </li>
</ul>
</div>  <!-- class_method_summary -->

<h2 class='h2_sum' id='instance_method_summary'>Instance Method Summary</h2>
<div class='div_sum'>
<ul class='summary full'>
  <li>
    <span class='summary_signature '>
      <a href="#==-instance_method" title="#== (instance method)">#<strong>==</strong>(obj)  &#x21d2; Boolean </a>
    </span>
    <div class='summary_desc'>
      <div class='inline'><p>Equality—If <em>obj</em> is not an <code>Exception</code>, returns <code>false</code>.</p></div>
    </div>
  </li>
  <li>
    <span class='summary_signature '>
      <a href="#backtrace-instance_method" title="#backtrace (instance method)">#<strong>backtrace</strong>  &#x21d2; Array<sup>?</sup> </a>
    </span>
    <div class='summary_desc'>
      <div class='inline'><p>Returns any backtrace associated with the exception.</p></div>
    </div>
  </li>
  <li>
    <span class='summary_signature '>
      <a href="#backtrace_locations-instance_method" title="#backtrace_locations (instance method)">#<strong>backtrace_locations</strong>  &#x21d2; Array<sup>?</sup> </a>
    </span>
    <div class='summary_desc'>
      <div class='inline'><p>Returns any backtrace associated with the exception.</p></div>
    </div>
  </li>
  <li>
    <span class='summary_signature '>
      <a href="#cause-instance_method" title="#cause (instance method)">#<strong>cause</strong>  &#x21d2; Exception<sup>?</sup> </a>
    </span>
    <div class='summary_desc'>
      <div class='inline'><p>Returns the previous exception ($!) at the time this exception was raised.</p></div>
    </div>
  </li>
  <li>
    <span class='summary_signature '>
      <a href="#detailed_message-instance_method" title="#detailed_message (instance method)">#<strong>detailed_message</strong>(highlight: bool, **opt)  &#x21d2; String </a>
    </span>
    <div class='summary_desc'>
      <div class='inline'><p>Processes a string returned by <a href="#message-instance_method" title="Exception#message (method)">#message</a>.</p></div>
    </div>
  </li>
  <li>
    <span class='summary_signature '>
      <a href="#exception-instance_method" title="#exception (instance method)">#<strong>exception</strong>([string])  &#x21d2; Exception </a>
    </span>
    <div class='summary_desc'>
      <div class='inline'><p>With no argument, or if the argument is the same as the receiver, return the receiver.</p></div>
    </div>
  </li>
  <li>
    <span class='summary_signature '>
      <a href="#full_message-instance_method" title="#full_message (instance method)">#<strong>full_message</strong>(highlight: bool, order: [:top or :bottom])  &#x21d2; String </a>
    </span>
    <div class='summary_desc'>
      <div class='inline'><p>Returns formatted string of <em>exception</em>.</p></div>
    </div>
  </li>
  <li>
    <span class='summary_signature '>
      <a href="#inspect-instance_method" title="#inspect (instance method)">#<strong>inspect</strong>  &#x21d2; String </a>
    </span>
    <div class='summary_desc'>
      <div class='inline'><p>Return this exception’s class name and message.</p></div>
    </div>
  </li>
  <li>
    <span class='summary_signature '>
      <a href="#message-instance_method" title="#message (instance method)">#<strong>message</strong>  &#x21d2; String </a>
    </span>
    <div class='summary_desc'>
      <div class='inline'><p>Returns the result of invoking <code>exception.to_s</code>.</p></div>
    </div>
  </li>
  <li>
    <span class='summary_signature '>
      <a href="#set_backtrace-instance_method" title="#set_backtrace (instance method)">#<strong>set_backtrace</strong>(backtrace)  &#x21d2; Array </a>
    </span>
    <div class='summary_desc'>
      <div class='inline'><p>Sets the backtrace information associated with <code>exc</code>.</p></div>
    </div>
  </li>
  <li>
    <span class='summary_signature '>
      <a href="#to_s-instance_method" title="#to_s (instance method)">#<strong>to_s</strong>  &#x21d2; String </a>
    </span>
    <div class='summary_desc'>
      <div class='inline'><p>Returns exception’s message (or the name of the exception if no message is set).</p></div>
    </div>
  </li>
</ul>
</div>  <!-- instance_method_summary -->
<h2 id='Constructor_Details' class='y_details'>Constructor Details</h2>

  <section class='method_details first' id="new-class_method">
  <h3 class='signature  first'>
    <div class='inline-block'>
      <span class='overload'>.<strong>new</strong>(msg = nil)  &#x21d2; <code>Exception</code> </span>
      <span class='overload'>.<strong>exception</strong>(msg = nil)  &#x21d2; <code>Exception</code> </span>
    </div>
  </h3>
<div class='docstring'>
  <div class='discussion'>
    
<p>Construct a new <code>Exception</code> object, optionally passing in a message.</p>

  </div>
</div>
<div class='tags'>
  
</div>
  <span class='link_repo'>[&#8202;<a class='repo' href='https://github.com/ruby/ruby/blob/v3_3_9/error.c#L1419-L1426'>GitHub</a>&#8202;]</span>
<div class='source_code h'>
  <pre class='lines_num' data-start='1419' data-end='1426'></pre>
  <div class='lines_code'>
<pre><span class='info file'># File 'error.c', line 1419</span></pre>
<pre class='code cpp'>

static VALUE
exc_initialize(int argc, VALUE *argv, VALUE exc)
{
    VALUE arg;

    arg = (!rb_check_arity(argc, 0, 1) ? Qnil : argv[0]);
    return exc_init(exc, arg);
}
</pre>
  </div>
</div>
</section>

<h2 class='y_details'>Class Attribute Details</h2>
<section class='method_details first' id="to_tty?-class_method">
  <h3 class='signature ro first'>
    .<strong>to_tty?</strong>  &#x21d2; <code>Boolean</code>  <span class="extras">(<span class='readonly'>readonly</span>)</span>  </h3>
<div class='docstring'>
  <div class='discussion'>
    
<p>Returns <code>true</code> if exception messages will be sent to a tty.</p>

  </div>
</div>
  <span class='link_repo'>[&#8202;<a class='repo' href='https://github.com/ruby/ruby/blob/v3_3_9/error.c#L1504-L1508'>GitHub</a>&#8202;]</span>
<div class='source_code h'>
  <pre class='lines_num' data-start='1504' data-end='1508'></pre>
  <div class='lines_code'>
<pre><span class='info file'># File 'error.c', line 1504</span></pre>
<pre class='code cpp'>

static VALUE
exc_s_to_tty_p(VALUE self)
{
    return RBOOL(rb_stderr_tty_p());
}
</pre>
  </div>
</div>
</section>

<h2 class='y_details'>Class Method Details</h2>
<section class='method_details first' id="exception-class_method">
  <h3 class='signature  first'>
    .<strong>exception</strong>([string])  &#x21d2; <code>Exception</code>   </h3>
<div class='docstring'>
  <div class='discussion'>
    
<p>With no argument, or if the argument is the same as the receiver, return the receiver. Otherwise, create a new exception object of the same class as the receiver, but with a message equal to <code>string.to_str</code>.</p>

  </div>
</div>
  <span class='link_repo'>[ <a class='repo' href='https://github.com/ruby/ruby/blob/v3_3_9/error.c#L3297'>GitHub</a> ]</span>

</section>

<h2 class='y_details'>Instance Method Details</h2>
<section class='method_details first' id="==-instance_method">
  <h3 class='signature  first'>
    #<strong>==</strong>(obj)  &#x21d2; <code>Boolean</code>   </h3>
<div class='docstring'>
  <div class='discussion'>
    
<p>Equality—If <em>obj</em> is not an <code>Exception</code>, returns <code>false</code>. Otherwise, returns <code>true</code> if <em>exc</em> and <em>obj</em> share same class, messages, and backtrace.</p>

  </div>
</div>
  <span class='link_repo'>[&#8202;<a class='repo' href='https://github.com/ruby/ruby/blob/v3_3_9/error.c#L1874-L1903'>GitHub</a>&#8202;]</span>
<div class='source_code h'>
  <pre class='lines_num' data-start='1874' data-end='1903'></pre>
  <div class='lines_code'>
<pre><span class='info file'># File 'error.c', line 1874</span></pre>
<pre class='code cpp'>

static VALUE
exc_equal(VALUE exc, VALUE obj)
{
    VALUE mesg, backtrace;

    if (exc == obj) return Qtrue;

    if (rb_obj_class(exc) != rb_obj_class(obj)) {
        int state;

        obj = rb_protect(try_convert_to_exception, obj, &amp;state);
        if (state || UNDEF_P(obj)) {
            rb_set_errinfo(Qnil);
            return Qfalse;
        }
        if (rb_obj_class(exc) != rb_obj_class(obj)) return Qfalse;
        mesg = rb_check_funcall(obj, id_message, 0, 0);
        if (UNDEF_P(mesg)) return Qfalse;
        backtrace = rb_check_funcall(obj, id_backtrace, 0, 0);
        if (UNDEF_P(backtrace)) return Qfalse;
    }
    else {
        mesg = rb_attr_get(obj, id_mesg);
        backtrace = exc_backtrace(obj);
    }

    if (!rb_equal(rb_attr_get(exc, id_mesg), mesg))
        return Qfalse;
    return rb_equal(exc_backtrace(exc), backtrace);
}
</pre>
  </div>
</div>
</section>

<section class='method_details' id="backtrace-instance_method">
  <h3 class='signature '>
    #<strong>backtrace</strong>  &#x21d2; <a href="Array.html" title="Array (class)">Array</a><sup>?</sup>   </h3>
<div class='docstring'>
  <div class='discussion'>
    
<p>Returns any backtrace associated with the exception. The backtrace is an array of strings, each containing either “filename:lineNo: in ‘method”’ or “filename:lineNo.”</p>

<pre class="code ruby"><code class="ruby"><span class='kw'>def</span> <span class='id identifier rubyid_a'>a</span>
  <span class='id identifier rubyid_raise'>raise</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>boom</span><span class='tstring_end'>&quot;</span></span>
<span class='kw'>end</span>

<span class='kw'>def</span> <span class='id identifier rubyid_b'>b</span>
  <span class='id identifier rubyid_a'>a</span>()
<span class='kw'>end</span>

<span class='kw'>begin</span>
  <span class='id identifier rubyid_b'>b</span>()
<span class='kw'>rescue</span> <span class='op'>=&gt;</span> <span class='id identifier rubyid_detail'>detail</span>
  <span class='id identifier rubyid_print'>print</span> <span class='id identifier rubyid_detail'>detail</span>.<span class='id identifier rubyid_backtrace'>backtrace</span>.<span class='id identifier rubyid_join'>join</span>(<span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\n</span><span class='tstring_end'>&quot;</span></span>)
<span class='kw'>end</span></code></pre>

<p><em>produces:</em></p>

<pre class="code ruby"><code class="ruby"><span class='id identifier rubyid_prog'>prog</span>.<span class='id identifier rubyid_rb'>rb</span><span class='symbeg'>:</span><span class='int'>2</span><span class='symbeg'>:</span><span class='kw'>in</span> <span class='backtick'>`</span><span class='tstring_content'>a&#39;
prog.rb:6:in </span><span class='tstring_end'>`</span></span><span class='id identifier rubyid_b'>b</span><span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>
prog.rb:10</span></code></pre>

<p>In the case no backtrace has been set, <code>nil</code> is returned</p>

<pre class="code ruby"><code class="ruby"><span class='id identifier rubyid_ex'>ex</span> <span class='op'>=</span> <span class='const'><a href="StandardError.html" title="StandardError (class)">StandardError</a></span>.<span class='id identifier rubyid_new'><a href="#new-class_method" title="Exception.new (method)">new</a></span>
<span class='id identifier rubyid_ex'>ex</span>.<span class='id identifier rubyid_backtrace'>backtrace</span>
<span class='comment'>#=&gt; nil</span></code></pre>

  </div>
</div>
  <span class='link_repo'>[&#8202;<a class='repo' href='https://github.com/ruby/ruby/blob/v3_3_9/error.c#L1740-L1753'>GitHub</a>&#8202;]</span>
<div class='source_code h'>
  <pre class='lines_num' data-start='1740' data-end='1753'></pre>
  <div class='lines_code'>
<pre><span class='info file'># File 'error.c', line 1740</span></pre>
<pre class='code cpp'>

static VALUE
exc_backtrace(VALUE exc)
{
    VALUE obj;

    obj = rb_attr_get(exc, id_bt);

    if (rb_backtrace_p(obj)) {
        obj = rb_backtrace_to_str_ary(obj);
        /* rb_ivar_set(exc, id_bt, obj); */
    }

    return obj;
}
</pre>
  </div>
</div>
</section>

<section class='method_details' id="backtrace_locations-instance_method">
  <h3 class='signature '>
    #<strong>backtrace_locations</strong>  &#x21d2; <a href="Array.html" title="Array (class)">Array</a><sup>?</sup>   </h3>
<div class='docstring'>
  <div class='discussion'>
    
<p>Returns any backtrace associated with the exception. This method is similar to <a href="#backtrace-instance_method" title="Exception#backtrace (method)">#backtrace</a>, but the backtrace is an array of <a href="Thread/Backtrace/Location.html" title="Thread::Backtrace::Location (class)"><code>::Thread::Backtrace::Location</code></a>.</p>

<p>This method is not affected by <a href="#set_backtrace-instance_method" title="Exception#set_backtrace (method)">#set_backtrace</a>().</p>

  </div>
</div>
  <span class='link_repo'>[&#8202;<a class='repo' href='https://github.com/ruby/ruby/blob/v3_3_9/error.c#L1788-L1798'>GitHub</a>&#8202;]</span>
<div class='source_code h'>
  <pre class='lines_num' data-start='1788' data-end='1798'></pre>
  <div class='lines_code'>
<pre><span class='info file'># File 'error.c', line 1788</span></pre>
<pre class='code cpp'>

static VALUE
exc_backtrace_locations(VALUE exc)
{
    VALUE obj;

    obj = rb_attr_get(exc, id_bt_locations);
    if (!NIL_P(obj)) {
        obj = rb_backtrace_to_location_ary(obj);
    }
    return obj;
}
</pre>
  </div>
</div>
</section>

<section class='method_details' id="cause-instance_method">
  <h3 class='signature '>
    #<strong>cause</strong>  &#x21d2; <code>Exception</code><sup>?</sup>   </h3>
<div class='docstring'>
  <div class='discussion'>
    
<p>Returns the previous exception ($!) at the time this exception was raised. This is useful for wrapping exceptions and retaining the original exception information.</p>

  </div>
</div>
  <span class='link_repo'>[&#8202;<a class='repo' href='https://github.com/ruby/ruby/blob/v3_3_9/error.c#L1853-L1857'>GitHub</a>&#8202;]</span>
<div class='source_code h'>
  <pre class='lines_num' data-start='1853' data-end='1857'></pre>
  <div class='lines_code'>
<pre><span class='info file'># File 'error.c', line 1853</span></pre>
<pre class='code cpp'>

static VALUE
exc_cause(VALUE exc)
{
    return rb_attr_get(exc, id_cause);
}
</pre>
  </div>
</div>
</section>

<section class='method_details' id="detailed_message-instance_method">
  <h3 class='signature '>
    #<strong>detailed_message</strong>(highlight: bool, **opt)  &#x21d2; <a href="String.html" title="String (class)">String</a>   </h3>
<div class='docstring'>
  <div class='discussion'>
    
<p>Processes a string returned by <a href="#message-instance_method" title="Exception#message (method)">#message</a>.</p>

<p>It may add the class name of the exception to the end of the first line. Also, when <code>highlight</code> keyword is true, it adds ANSI escape sequences to make the message bold.</p>

<p>If you override this method, it must be tolerant for unknown keyword arguments. All keyword arguments passed to <a href="#full_message-instance_method" title="Exception#full_message (method)">#full_message</a> are delegated to this method.</p>

<p>This method is overridden by did_you_mean and error_highlight to add their information.</p>

<p>A user-defined exception class can also define their own <code>detailed_message</code> method to add supplemental information. When <code>highlight</code> is true, it can return a string containing escape sequences, but use widely-supported ones. It is recommended to limit the following codes:</p>
<ul><li>
<p>Reset (<code>\e[0m</code>)</p>
</li><li>
<p>Bold (<code>\e[1m</code>)</p>
</li><li>
<p>Underline (<code>\e[4m</code>)</p>
</li><li>
<p>Foreground color except white and black</p>
<ul><li>
<p>Red (<code>\e[31m</code>)</p>
</li><li>
<p>Green (<code>\e[32m</code>)</p>
</li><li>
<p>Yellow (<code>\e[33m</code>)</p>
</li><li>
<p>Blue (<code>\e[34m</code>)</p>
</li><li>
<p>Magenta (<code>\e[35m</code>)</p>
</li><li>
<p>Cyan (<code>\e[36m</code>)</p>
</li></ul>
</li></ul>

<p>Use escape sequences carefully even if <code>highlight</code> is true. Do not use escape sequences to express essential information; the message should be readable even if all escape sequences are ignored.</p>

  </div>
</div>
  <span class='link_repo'>[&#8202;<a class='repo' href='https://github.com/ruby/ruby/blob/v3_3_9/error.c#L1656-L1668'>GitHub</a>&#8202;]</span>
<div class='source_code h'>
  <pre class='lines_num' data-start='1656' data-end='1668'></pre>
  <div class='lines_code'>
<pre><span class='info file'># File 'error.c', line 1656</span></pre>
<pre class='code cpp'>

static VALUE
exc_detailed_message(int argc, VALUE *argv, VALUE exc)
{
    VALUE opt;

    rb_scan_args(argc, argv, &quot;0:&quot;, &amp;opt);

    VALUE highlight = check_highlight_keyword(opt, 0);

    extern VALUE rb_decorate_message(const VALUE eclass, VALUE emesg, int highlight);

    return rb_decorate_message(CLASS_OF(exc), rb_get_message(exc), RTEST(highlight));
}
</pre>
  </div>
</div>
</section>

<section class='method_details' id="exception-instance_method">
  <h3 class='signature '>
    #<strong>exception</strong>([string])  &#x21d2; <code>Exception</code>   </h3>
<div class='docstring'>
  <div class='discussion'>
    
<p>With no argument, or if the argument is the same as the receiver, return the receiver. Otherwise, create a new exception object of the same class as the receiver, but with a message equal to <code>string.to_str</code>.</p>

  </div>
</div>
  <span class='link_repo'>[&#8202;<a class='repo' href='https://github.com/ruby/ruby/blob/v3_3_9/error.c#L1441-L1452'>GitHub</a>&#8202;]</span>
<div class='source_code h'>
  <pre class='lines_num' data-start='1441' data-end='1452'></pre>
  <div class='lines_code'>
<pre><span class='info file'># File 'error.c', line 1441</span></pre>
<pre class='code cpp'>

static VALUE
exc_exception(int argc, VALUE *argv, VALUE self)
{
    VALUE exc;

    argc = rb_check_arity(argc, 0, 1);
    if (argc == 0) return self;
    if (argc == 1 &amp;&amp; self == argv[0]) return self;
    exc = rb_obj_clone(self);
    rb_ivar_set(exc, id_mesg, argv[0]);
    return exc;
}
</pre>
  </div>
</div>
</section>

<section class='method_details' id="full_message-instance_method">
  <h3 class='signature '>
    #<strong>full_message</strong>(highlight: bool, order: [:top or :bottom])  &#x21d2; <a href="String.html" title="String (class)">String</a>   </h3>
<div class='docstring'>
  <div class='discussion'>
    
<p>Returns formatted string of <em>exception</em>. The returned string is formatted using the same format that Ruby uses when printing an uncaught exceptions to stderr.</p>

<p>If <em>highlight</em> is <code>true</code> the default error handler will send the messages to a tty.</p>

<p><em>order</em> must be either of <code>:top</code> or <code>:bottom</code>, and places the error message and the innermost backtrace come at the top or the bottom.</p>

<p>The default values of these options depend on <code>$stderr</code> and its <code>tty?</code> at the timing of a call.</p>

  </div>
</div>
  <span class='link_repo'>[&#8202;<a class='repo' href='https://github.com/ruby/ruby/blob/v3_3_9/error.c#L1578-L1600'>GitHub</a>&#8202;]</span>
<div class='source_code h'>
  <pre class='lines_num' data-start='1578' data-end='1600'></pre>
  <div class='lines_code'>
<pre><span class='info file'># File 'error.c', line 1578</span></pre>
<pre class='code cpp'>

static VALUE
exc_full_message(int argc, VALUE *argv, VALUE exc)
{
    VALUE opt, str, emesg, errat;
    VALUE highlight, order;

    rb_scan_args(argc, argv, &quot;0:&quot;, &amp;opt);

    highlight = check_highlight_keyword(opt, 1);
    order = check_order_keyword(opt);

    {
        if (NIL_P(opt)) opt = rb_hash_new();
        rb_hash_aset(opt, sym_highlight, highlight);
    }

    str = rb_str_new2(&quot;&quot;);
    errat = rb_get_backtrace(exc);
    emesg = rb_get_detailed_message(exc, opt);

    rb_error_write(exc, emesg, errat, str, opt, highlight, order);
    return str;
}
</pre>
  </div>
</div>
</section>

<section class='method_details' id="inspect-instance_method">
  <h3 class='signature '>
    #<strong>inspect</strong>  &#x21d2; <a href="String.html" title="String (class)">String</a>   </h3>
<div class='docstring'>
  <div class='discussion'>
    
<p>Return this exception’s class name and message.</p>

  </div>
</div>
  <span class='link_repo'>[&#8202;<a class='repo' href='https://github.com/ruby/ruby/blob/v3_3_9/error.c#L1677-L1703'>GitHub</a>&#8202;]</span>
<div class='source_code h'>
  <pre class='lines_num' data-start='1677' data-end='1703'></pre>
  <div class='lines_code'>
<pre><span class='info file'># File 'error.c', line 1677</span></pre>
<pre class='code cpp'>

static VALUE
exc_inspect(VALUE exc)
{
    VALUE str, klass;

    klass = CLASS_OF(exc);
    exc = rb_obj_as_string(exc);
    if (RSTRING_LEN(exc) == 0) {
        return rb_class_name(klass);
    }

    str = rb_str_buf_new2(&quot;#&lt;&quot;);
    klass = rb_class_name(klass);
    rb_str_buf_append(str, klass);

    if (RTEST(rb_str_include(exc, rb_str_new2(&quot;\n&quot;)))) {
        rb_str_catf(str, &quot;:%+&quot;PRIsVALUE, exc);
    }
    else {
        rb_str_buf_cat(str, &quot;: &quot;, 2);
        rb_str_buf_append(str, exc);
    }

    rb_str_buf_cat(str, &quot;&gt;&quot;, 1);

    return str;
}
</pre>
  </div>
</div>
</section>

<section class='method_details' id="message-instance_method">
  <h3 class='signature '>
    #<strong>message</strong>  &#x21d2; <a href="String.html" title="String (class)">String</a>   </h3>
<div class='docstring'>
  <div class='discussion'>
    
<p>Returns the result of invoking <code>exception.to_s</code>. Normally this returns the exception’s message or name.</p>

  </div>
</div>
  <span class='link_repo'>[&#8202;<a class='repo' href='https://github.com/ruby/ruby/blob/v3_3_9/error.c#L1610-L1614'>GitHub</a>&#8202;]</span>
<div class='source_code h'>
  <pre class='lines_num' data-start='1610' data-end='1614'></pre>
  <div class='lines_code'>
<pre><span class='info file'># File 'error.c', line 1610</span></pre>
<pre class='code cpp'>

static VALUE
exc_message(VALUE exc)
{
    return rb_funcallv(exc, idTo_s, 0, 0);
}
</pre>
  </div>
</div>
</section>

<section class='method_details' id="set_backtrace-instance_method">
  <h3 class='signature '>
    #<strong>set_backtrace</strong>(backtrace)  &#x21d2; <a href="Array.html" title="Array (class)">Array</a>   </h3>
<div class='docstring'>
  <div class='discussion'>
    
<p>Sets the backtrace information associated with <code>exc</code>. The <a href="#backtrace-instance_method" title="Exception#backtrace (method)">#backtrace</a> must be an array of <a href="String.html" title="String (class)"><code>::String</code></a> objects or a single <a href="String.html" title="String (class)"><code>::String</code></a> in the format described in <a href="#backtrace-instance_method" title="Exception#backtrace (method)">#backtrace</a>.</p>

  </div>
</div>
  <span class='link_repo'>[&#8202;<a class='repo' href='https://github.com/ruby/ruby/blob/v3_3_9/error.c#L1832-L1836'>GitHub</a>&#8202;]</span>
<div class='source_code h'>
  <pre class='lines_num' data-start='1832' data-end='1836'></pre>
  <div class='lines_code'>
<pre><span class='info file'># File 'error.c', line 1832</span></pre>
<pre class='code cpp'>

static VALUE
exc_set_backtrace(VALUE exc, VALUE bt)
{
    return rb_ivar_set(exc, id_bt, rb_check_backtrace(bt));
}
</pre>
  </div>
</div>
</section>

<section class='method_details' id="to_s-instance_method">
  <h3 class='signature '>
    #<strong>to_s</strong>  &#x21d2; <a href="String.html" title="String (class)">String</a>   </h3>
<div class='docstring'>
  <div class='discussion'>
    
<p>Returns exception’s message (or the name of the exception if no message is set).</p>

  </div>
</div>
  <span class='link_repo'>[&#8202;<a class='repo' href='https://github.com/ruby/ruby/blob/v3_3_9/error.c#L1462-L1469'>GitHub</a>&#8202;]</span>
<div class='source_code h'>
  <pre class='lines_num' data-start='1462' data-end='1469'></pre>
  <div class='lines_code'>
<pre><span class='info file'># File 'error.c', line 1462</span></pre>
<pre class='code cpp'>

static VALUE
exc_to_s(VALUE exc)
{
    VALUE mesg = rb_attr_get(exc, idMesg);

    if (NIL_P(mesg)) return rb_class_name(CLASS_OF(exc));
    return rb_String(mesg);
}
</pre>
  </div>
</div>
</section>


<div id='footer'></div>
</div> <!-- content  -->
</div> <!-- y_main   -->
</body>
</html>